File: base/tps/shared/webapps/tps/js/selftest.js
/* --- BEGIN COPYRIGHT BLOCK ---
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Copyright (C) 2013 Red Hat, Inc.
 * All rights reserved.
 * --- END COPYRIGHT BLOCK ---
 *
 * @author Endi S. Dewata
 */

var SelfTestModel = Model.extend({
    urlRoot: "/tps/rest/selftests",
    parseResponse: function(response) {
        return {
            id: response.id,
            enabledAtStartup: response.EnabledAtStartup,
            criticalAtStartup: response.CriticalAtStartup,
            enabledOnDemand: response.EnabledOnDemand,
            criticalOnDemand: response.CriticalOnDemand,
        };
    },
    createRequest: function(attributes) {
        return {
            id: attributes.id,
            EnabledAtStartup: attributes.enabledAtStartup,
            CriticalAtStartup: attributes.criticalAtStartup,
            EnabledOnDemand: attributes.enabledOnDemand,
            CriticalOnDemand: attributes.criticalOnDemand
        };
    },
    run: function(options) {
        var self = this;
        $.ajax({
            type: "POST",
            url: self.url() + "/run",
            dataType: "json"
        }).done(function(data, textStatus, jqXHR) {
            if (options.success) options.success.call(self, data, textStatus, jqXHR);
        }).fail(function(jqXHR, textStatus, errorThrown) {
            if (options.error) options.error.call(self, jqXHR, textStatus, errorThrown);
        });
    }
});

var SelfTestCollection = Collection.extend({
    urlRoot: "/tps/rest/selftests",
    getEntries: function(response) {
        return response.entries;
    },
    getLinks: function(response) {
        return response.Link;
    },
    parseEntry: function(entry) {
        return new SelfTestModel({
            id: entry.id,
            enabledAtStartup: entry.EnabledAtStartup,
            criticalAtStartup: entry.CriticalAtStartup,
            enabledOnDemand: entry.EnabledOnDemand,
            criticalOnDemand: entry.CriticalOnDemand,
        });
    }
});

var SelfTestPage = EntryPage.extend({
    initialize: function(options) {
        var self = this;
        SelfTestPage.__super__.initialize.call(self, options);
    },
    setup: function() {
        var self = this;

        SelfTestPage.__super__.setup.call(self);

        self.runAction = $("[name='run']", self.viewMenu);

        $("a", self.runAction).click(function(e) {

            e.preventDefault();

            self.model.run({
                success: function(data, textStatus, jqXHR) {
                    self.showResult({
                        id: data.id,
                        status: data.Status,
                        output: data.Output
                    });
                },
                error: function(jqXHR, textStatus, errorThrown) {
                    self.showResult({
                        id: self.model.get("id"),
                        status: textStatus,
                        output: errorThrown
                    });
                }
            });

        });
    },
    showResult: function(data) {
        var dialog = new Dialog({
            el: self.$("#selftest-result-dialog"),
            title: "Self Test Result",
            readonly: ["id", "status", "output"],
            actions: ["close"]
        });

        dialog.entry = data;

        dialog.open();
    }
});

var SelfTestsTable = ModelTable.extend({
    initialize: function(options) {
        var self = this;
        SelfTestsTable.__super__.initialize.call(self, options);

        self.runButton = $("[name='run']", self.buttons);
        self.runButton.click(function(e) {
            var items = self.getSelectedRows();
            _.each(items, function(item, index) {
                self.runTest(item, index);
            });
        });

        self.clearButton = $("[name='clear']", self.buttons);
        self.clearButton.click(function(e) {
            var items = self.getSelectedRows();
            _.each(items, function(item, index) {
                self.clearTest(item, index);
            });
        });
    },
    runTest: function(item, index) {
        var self = this;

        var statusTD = $("td[name='status']", item.$el);
        statusTD.text("RUNNING");

        var id = item.get("id");
        var model = self.collection.get(id);

        model.run({
            success: function(data, textStatus, jqXHR) {
                statusTD.empty();
                var link = $("<a/>", {
                    text: data.Status,
                    click: function(e) {
                        e.preventDefault();
                        self.showResult({
                            id: data.id,
                            status: data.Status,
                            output: data.Output
                        });
                    }
                }).appendTo(statusTD);
            },
            error: function(jqXHR, textStatus, errorThrown) {
                statusTD.empty();
                var link = $("<a/>", {
                    text: textStatus,
                    click: function(e) {
                        e.preventDefault();
                        self.showResult({
                            id: id,
                            status: textStatus,
                            output: errorThrown
                        });
                    }
                }).appendTo(statusTD);
            }
        });
    },
    clearTest: function(item, index) {
        var self = this;

        var statusTD = $("td[name='status']", item.$el);
        statusTD.empty();
    },
    showResult: function(data) {
        var dialog = new Dialog({
            el: self.parent.$("#selftest-result-dialog"),
            title: "Self Test Result",
            readonly: ["id", "status", "output"],
            actions: ["close"]
        });

        dialog.entry = data;

        dialog.open();
    }
});

var SelfTestsPage = Page.extend({
    load: function() {
        var self = this;

        var table = new SelfTestsTable({
            el: self.$("table[name='selftests']"),
            collection: new SelfTestCollection(),
            parent: self
        });

        table.render();
    }
});

File: base/server/python/pki/server/cli/selftest.py
# Authors:
#     Dinesh Prasanth M K <dmoluguw@redhat.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 2 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Copyright (C) 2015-2017 Red Hat, Inc.
# All rights reserved.
#

import argparse
import sys
import logging

import pki.cli
import pki.server


class SelfTestCLI(pki.cli.CLI):
    def __init__(self):
        super().__init__('selftest', 'Selftest management commands')
        self.add_module(EnableSelfTestCLI())
        self.add_module(DisableSelftestCLI())


class EnableSelfTestCLI(pki.cli.CLI):
    def __init__(self):
        super().__init__('enable', 'Enable selftests.')

        self.parser = argparse.ArgumentParser(
            prog=self.name,
            add_help=False)
        self.parser.add_argument(
            '-i',
            '--instance',
            default='pki-tomcat')
        self.parser.add_argument(
            '--subsystem',
            action='append')
        self.parser.add_argument(
            '-v',
            '--verbose',
            action='store_true')
        self.parser.add_argument(
            '--debug',
            action='store_true')
        self.parser.add_argument(
            '--help',
            action='store_true')
        self.parser.add_argument(
            'selftest_id',
            nargs='?')

    def print_help(self):
        print('Usage: pki-server selftest-enable [OPTIONS] [<Selftest ID>]')
        print()
        print('  -i, --instance <instance ID>    Instance ID (default: pki-tomcat).')
        print('  --subsystem <subsystem name>    Subsystem Name.')
        print('  --help                          Show help message.')
        print()

    def execute(self, argv, args=None):

        if not args:
            args = self.parser.parse_args(args=argv)

        if args.help:
            self.print_help()
            return

        if args.debug:
            logging.getLogger().setLevel(logging.DEBUG)

        elif args.verbose:
            logging.getLogger().setLevel(logging.INFO)

        instance_name = args.instance
        subsystem_names = args.subsystem
        test = args.selftest_id

        # Load instance
        instance = pki.server.PKIServerFactory.create(instance_name)

        if not instance.exists():
            print('ERROR: Invalid instance %s.' % instance_name)
            sys.exit(1)

        instance.load()

        # To hold the instance of the loaded subsystems
        target_subsystems = []

        # Load subsystem or subsystems
        if not subsystem_names:
            for subsys in instance.get_subsystems():
                target_subsystems.append(subsys)
        else:
            for subsys in subsystem_names:
                target_subsystems.append(instance.get_subsystem(subsys))

        try:
            # Enable critical tests for all subsystems listed in target_subsystems
            for subsys in target_subsystems:
                subsys.set_startup_test_criticality(test=test, critical=True)
                # Save the updated CS.cfg to disk
                subsys.save()

        except pki.server.PKIServerException as e:
            logging.error(str(e))
            sys.exit(1)


class DisableSelftestCLI(pki.cli.CLI):
    def __init__(self):
        super().__init__('disable', 'Disable selftests.')

        self.parser = argparse.ArgumentParser(
            prog=self.name,
            add_help=False)
        self.parser.add_argument(
            '-i',
            '--instance',
            default='pki-tomcat')
        self.parser.add_argument(
            '--subsystem',
            action='append')
        self.parser.add_argument(
            '-v',
            '--verbose',
            action='store_true')
        self.parser.add_argument(
            '--debug',
            action='store_true')
        self.parser.add_argument(
            '--help',
            action='store_true')
        self.parser.add_argument(
            'selftest_id',
            nargs='?')

    def print_help(self):
        print('Usage: pki-server selftest-disable [OPTIONS] [<Selftest ID>]')
        print()
        print('  -i, --instance <instance ID>    Instance ID (default: pki-tomcat).')
        print('  --subsystem <subsystem name>    Subsystem Name.')
        print('  --help                          Show help message.')
        print()

    def execute(self, argv, args=None):

        if not args:
            args = self.parser.parse_args(args=argv)

        if args.help:
            self.print_help()
            return

        if args.debug:
            logging.getLogger().setLevel(logging.DEBUG)

        elif args.verbose:
            logging.getLogger().setLevel(logging.INFO)

        instance_name = args.instance
        subsystem_names = args.subsystem
        test = args.selftest_id

        # Load instance
        instance = pki.server.PKIServerFactory.create(instance_name)

        if not instance.exists():
            print('ERROR: Invalid instance %s.' % instance_name)
            sys.exit(1)

        instance.load()

        # To hold the instance of the loaded subsystems
        target_subsystems = []

        # Load subsystem or subsystems
        if not subsystem_names:
            for subsys in instance.get_subsystems():
                target_subsystems.append(subsys)
        else:
            for subsys in subsystem_names:
                target_subsystems.append(instance.get_subsystem(subsys))

        try:
            # Disable critical tests for all subsystems listed in target_subsystems
            for subsys in target_subsystems:
                subsys.set_startup_test_criticality(test=test, critical=False)
                # Save the updated CS.cfg to disk
                subsys.save()

        except pki.server.PKIServerException as e:
            logging.error(str(e))
            sys.exit(1)
